sycl/test-e2e/TriStateExpr.py

```diff
@@ -0,0 +1,176 @@
+from lit.BooleanExpression import BooleanExpression
+
+
+class E2EExpr(BooleanExpression):
+    build_specific_features = {
+        "build-and-run-mode",
+        "target-spir",
+        "target-nvidia",
+        "target-amd",
+        "native_cpu",
+        "linux",
+        "system-linux",
+        "windows",
+        "system-windows",
+        "enable-perf-tests",
+        "opencl_icd",
+        "cuda_dev_kit",
+        "zstd",
+        "vulkan",
+        "true",
+        "false",
+    }
+
+    def __init__(self, string, variables, build_only_mode, ignore_value):
+        BooleanExpression.__init__(self, string, variables)
+        self.build_only_mode = build_only_mode
+        self.ignore = False
+        self.ignore_value = ignore_value
+
+    @staticmethod
+    def evaluate(string, variables, build_only_mode, ignore_value=True):
+        try:
+            parser = E2EExpr(string, set(variables), build_only_mode, ignore_value)
+            return parser.parseAll()
+        except ValueError as e:
+            raise ValueError(str(e) + ("\nin expression: %r" % string))
+
+    def parseMATCH(self):
+        token = self.token
+        BooleanExpression.parseMATCH(self)
+        if token not in self.build_specific_features and self.build_only_mode:
+            self.ignore = True
+        else:
+            self.ignore = False
+
+    def parseAND(self):
+        self.parseNOT()
+        while self.accept("&&"):
+            left = self.value
+            left_ignore = self.ignore
+            self.parseNOT()
+            right = self.value
+            right_ignore = self.ignore
+            self.value = left and right
+            # Ignore if both are ignore or if one is true and the other is ignore
+            self.ignore = (
+                (left_ignore and right_ignore)
+                or (left_ignore and right)
+                or (left and right_ignore)
+            )
+
+    def parseOR(self):
+        self.parseAND()
+        while self.accept("||"):
+            left = self.value
+            left_ignore = self.ignore
+            self.parseAND()
+            right = self.value
+            right_ignore = self.ignore
+            self.value = left or right
+            # Ignore if both are ignore or if one is false and the other is ignore
```

Comment:
If combined with the above `unsatisfied_build_feat || (runtime_feat_A && !runtime_feat_A)` would result in ignore. This is still contrived, but I'm not sure you can not end up with invalid `ignore`s given that the invariant of "Expression result is 'ignore' if and only if the ignored variables can always be set in such a way to make the result (when the expression is evaluated as a normal Boolean expression) either false or true"

Reply:
I think there are two viewpoints here. One, purely formal/math-like, another having extra semantics of "ignore runtime properties/runtime context". Having "build_property || runtime_property" would mean that the build isn't actually broken when "runtime_property" is `true`, so it is actually using "build_property" as a run-time condition (in run-time context might be a better wording), so it will be fine to ignore in build only mode.

I think the same would apply to your other comment.

Reply:
I guess that's right, resolving this for now.